customer_outstanding_statement/report/customer_outstanding_statement.py

```diff
@@ -0,0 +1,319 @@
+            COALESCE(Q2.currency_id, c.currency_id) AS currency_id, open_amount
+            FROM Q2
+            JOIN res_company c ON (c.id = Q2.company_id)
+            WHERE c.id = %s
+        """ % company_id
+
+    def _get_account_display_lines(self, company_id, partner_ids, date_end):
+        res = dict(map(lambda x: (x, []), partner_ids))
+        partners = ', '.join([str(i) for i in partner_ids])
+        date_end = datetime.strptime(
+            date_end, DEFAULT_SERVER_DATE_FORMAT).date()
+        self.env.cr.execute("""WITH Q1 AS (%s), Q2 AS (%s), Q3 AS (%s)
+        SELECT partner_id, currency_id, move_id, date, date_maturity, debit,
+                            credit, amount, open_amount, name, ref, blocked
+        FROM Q3
+        ORDER BY date, date_maturity, move_id""" % (
+            self._display_lines_sql_q1(partners, date_end),
+            self._display_lines_sql_q2(),
+            self._display_lines_sql_q3(company_id)))
+        for row in self.env.cr.dictfetchall():
+            res[row.pop('partner_id')].append(row)
+        return res
+
+    def _show_buckets_sql_q1(self, partners, date_end):
+        return """
+            SELECT l.partner_id, l.currency_id, l.company_id, l.move_id,
+            CASE WHEN l.balance > 0.0
+                THEN l.balance - sum(coalesce(pd.amount, 0.0))
+                ELSE l.balance + sum(coalesce(pc.amount, 0.0))
+            END AS open_due,
+            CASE WHEN l.balance > 0.0
+                THEN l.amount_currency - sum(coalesce(pd.amount_currency, 0.0))
+                ELSE l.amount_currency + sum(coalesce(pc.amount_currency, 0.0))
+            END AS open_due_currency,
+            CASE WHEN l.date_maturity is null
+                THEN l.date
+                ELSE l.date_maturity
+            END as date_maturity
+            FROM account_move_line l
+            JOIN account_account_type at ON (at.id = l.user_type_id)
+            JOIN account_move m ON (l.move_id = m.id)
+            LEFT JOIN (SELECT pr.*
+                FROM account_partial_reconcile pr
+                INNER JOIN account_move_line l2
+                ON pr.credit_move_id = l2.id
+                WHERE l2.date <= '%s'
+            ) as pd ON pd.debit_move_id = l.id
+            LEFT JOIN (SELECT pr.*
+                FROM account_partial_reconcile pr
+                INNER JOIN account_move_line l2
+                ON pr.debit_move_id = l2.id
+                WHERE l2.date <= '%s'
+            ) as pc ON pc.credit_move_id = l.id
+            WHERE l.partner_id IN (%s) AND at.type = 'receivable'
+                                AND not l.reconciled AND not l.blocked
+            GROUP BY l.partner_id, l.currency_id, l.date, l.date_maturity,
+                                l.amount_currency, l.balance, l.move_id,
+                                l.company_id
+        """ % (date_end, date_end, partners)
+
+    def _show_buckets_sql_q2(self, today, minus_30, minus_60, minus_90,
```

Comment:
@hveficent For consistency, i think this one should be printed with today as of date_end from the wizard...

Reply:
@feketemihai No, the buckets are something aside from the proper report. The report is linked to a specific date (in this case, the date_end). But the buckets (the aging report) are not linked to this date, expressly. They show what is due or not due of today and several months before. See to how they are defined in other websites:
- https://fitsmallbusiness.com/accounts-receivable-aging-report-quickbooks-online/
- https://www.thebalance.com/accounts-receivable-aging-report-397853
- https://www.accountingcoach.com/blog/what-is-the-aging-method

Reply:
@mreficent I agree with you, but the is printed on date_end, and probably will be 2 amounts showed, so the customer will not know what it represent, i  am thinking since you print it at specific date, moves after that date not to be considered...

Reply:
If you don't want the customers to get confused, then don't mark the buckets option or explain to them want the buckets and the proper report means. 🤔